cmd/juju/application/showunit.go

```diff
@@ -0,0 +1,283 @@
+	if c.relatedUnit != "" && !names.IsValidUnit(c.relatedUnit) {
+		return errors.NotValidf("related unit name %v", c.relatedUnit)
+	}
+	var invalid []string
+	for _, one := range c.units {
+		if !names.IsValidUnit(one) {
+			invalid = append(invalid, one)
+		}
+	}
+	if len(invalid) == 0 {
+		return nil
+	}
+	plural := "s"
+	if len(invalid) == 1 {
+		plural = ""
+	}
+	return errors.NotValidf(`unit name%v %v`, plural, strings.Join(invalid, `, `))
+}
+
+// SetFlags implements Command.SetFlags.
+func (c *showUnitCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	c.out.AddFlags(f, "yaml", cmd.DefaultFormatters.Formatters())
+	f.StringVar(&c.endpoint, "endpoint", "", "Only show relation data for the specified endpoint")
+	f.StringVar(&c.relatedUnit, "related-unit", "", "Only show relation data for the specified unit")
+	f.BoolVar(&c.appOnly, "app", false, "Only show application relation data")
+}
+
+// UnitsInfoAPI defines the API methods that show-unit command uses.
+type UnitsInfoAPI interface {
+	Close() error
+	BestAPIVersion() int
+	UnitsInfo([]names.UnitTag) ([]params.UnitInfoResult, error)
+}
+
+func (c *showUnitCommand) newUnitAPI() (UnitsInfoAPI, error) {
+	root, err := c.NewAPIRoot()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return application.NewClient(root), nil
+}
+
+// Info implements Command.Run.
+func (c *showUnitCommand) Run(ctx *cmd.Context) error {
+	client, err := c.newAPIFunc()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	if v := client.BestAPIVersion(); v < 12 {
+		// old client does not support showing applications.
+		return errors.NotSupportedf("show unit on API server version %v", v)
+	}
+
+	tags, err := c.getUnitTags()
+	if err != nil {
+		return err
+	}
+
+	results, err := client.UnitsInfo(tags)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var errs params.ErrorResults
+	var valid []params.UnitResult
+	for _, result := range results {
+		if result.Error != nil {
+			errs.Results = append(errs.Results, params.ErrorResult{result.Error})
+			continue
+		}
+		valid = append(valid, *result.Result)
+	}
+	if len(errs.Results) > 0 {
+		return errs.Combine()
+	}
+
+	output, err := c.formatUnitInfos(valid)
+	if err != nil {
+		return err
+	}
+	return c.out.Write(ctx, output)
+}
+
+func (c *showUnitCommand) getUnitTags() ([]names.UnitTag, error) {
+	tags := make([]names.UnitTag, len(c.units))
+	for i, one := range c.units {
+		if !names.IsValidUnit(one) {
+			return nil, errors.Errorf("invalid unit name %v", one)
+		}
+		tags[i] = names.NewUnitTag(one)
+	}
+	return tags, nil
+}
+
+func (c *showUnitCommand) formatUnitInfos(all []params.UnitResult) (map[string]UnitInfo, error) {
+	if len(all) == 0 {
+		return nil, nil
+	}
+	output := make(map[string]UnitInfo)
+	for _, one := range all {
+		tag, info, err := c.createUnitInfo(one)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		output[tag.Id()] = info
+	}
+	return output, nil
+}
+
+type UnitRelationData struct {
```

Comment:
These types could be in the API package and returned from the API itself.

Reply:
These have (UX related) format tags and are structured to be fit for purpose for the layer in which they are used, ie the CLI. The API layer should have its own structs, even if similar. Each layer transition should use different structs